lib/modules/manager/pipenv/artifacts.ts

```diff
@@ -24,38 +26,83 @@ import type {
 import { extractPackageFile } from './extract';
 import { PipfileLockSchema } from './schema';
 
-export function getPythonConstraint(
+export async function getPythonConstraint(
+  pipfileName: string,
+  pipfileContent: string,
   existingLockFileContent: string,
   config: UpdateArtifactsConfig,
-): string | undefined {
+): Promise<string | undefined> {
   const { constraints = {} } = config;
   const { python } = constraints;
 
   if (python) {
-    logger.debug('Using python constraint from config');
+    logger.debug(`Using python constraint ${python} from config`);
     return python;
   }
+
+  // Try Pipfile first because it may have had its Python version updated
+  try {
+    const pipfile = parseToml(pipfileContent) as any;
```

Comment:
Please create a minimal schema instead of any 😏

Reply:
I don't have time right now, and it's missing in extract. Once extract is done we can reuse the same logic